### ISSUE TITLE
Dataset create call failed in LightGBM with error: bad allocation

### ISSUE THREAD

**Issue by JWenBin** [Azure/mmlspark#783]
Sometimes my code runs right while sometimes with "error: bad allocation". But the code is the same code. 
When it runs failed：
Exception in thread "main" org.apache.spark.SparkException: Job aborted due to stage failure: Could not recover from a failed barrier ResultStage. Most recent failure reason: Stage failed because barrier task ResultTask(7, 0) finished unsuccessfully.
java.lang.Exception: Dataset create call failed in LightGBM with error: bad allocation
	at com.microsoft.ml.spark.lightgbm.LightGBMUtils$.validate(LightGBMUtils.scala:31)
	at com.microsoft.ml.spark.lightgbm.LightGBMUtils$.generateSparseDataset(LightGBMUtils.scala:246)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$.generateDataset(TrainUtils.scala:79)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$.translate(TrainUtils.scala:258)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$.trainLightGBM(TrainUtils.scala:411)
	at com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6.apply(LightGBMBase.scala:85)
	at com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6.apply(LightGBMBase.scala:85)
	at org.apache.spark.rdd.RDDBarrier$$anonfun$mapPartitions$1$$anonfun$apply$1.apply(RDDBarrier.scala:51)
	at org.apache.spark.rdd.RDDBarrier$$anonfun$mapPartitions$1$$anonfun$apply$1.apply(RDDBarrier.scala:51)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:402)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:408)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1887)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1875)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1874)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1874)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskCompletion(DAGScheduler.scala:1663)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2105)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2057)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2046)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2158)
	at org.apache.spark.rdd.RDD$$anonfun$reduce$1.apply(RDD.scala:1035)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
	at org.apache.spark.rdd.RDD.reduce(RDD.scala:1017)
	at com.microsoft.ml.spark.lightgbm.LightGBMBase$class.innerTrain(LightGBMBase.scala:88)
	at com.microsoft.ml.spark.lightgbm.LightGBMClassifier.innerTrain(LightGBMClassifier.scala:26)
	at com.microsoft.ml.spark.lightgbm.LightGBMBase$class.train(LightGBMBase.scala:38)
	at com.microsoft.ml.spark.lightgbm.LightGBMClassifier.train(LightGBMClassifier.scala:26)
	at com.microsoft.ml.spark.lightgbm.LightGBMClassifier.train(LightGBMClassifier.scala:26)
	at org.apache.spark.ml.Predictor.fit(Predictor.scala:118)
	at main.Lgbm$.main(Lgbm.scala:56)
	at main.Lgbm.main(Lgbm.scala)

When it runs without error：
LightGBM] [Warning] Starting from the 2.1.2 version, default value for the "boost_from_average" parameter in "binary" objective is true.
This may cause significantly different results comparing to the previous versions of LightGBM.
Try to set boost_from_average=false, if your old models produce bad results
[LightGBM] [Info] Number of positive: 31, number of negative: 169
[LightGBM] [Info] Total Bins 29
[LightGBM] [Info] Number of data: 200, number of used features: 16
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.155000 -> initscore=-1.695912
[LightGBM] [Info] Start training from score -1.695912
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
......
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
+------+------------+
|target|predictLabel|
+------+------------+
|1     |0.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
|1     |1.0         |
|0     |0.0         |
|1     |1.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
|1     |1.0         |
|0     |0.0         |
|0     |0.0         |
|0     |0.0         |
+------+------------+
only showing top 20 rows

**Comment by JWenBin** [comment]
Here is my code：
	val df =spark.read.load("trainData")
	  .selectExpr("target","deviceid","newsid","guid","pos","app_version","device_vendor","netmodel","osversion","lng","lat","device_version","ts")
	  .limit(200)
	  .na.fill("str")
	  .cache()
	val fdf: PipelineModel =LightGBMUtils.featurizeData(df,"target","features")
	val df2=fdf.transform(df).selectExpr("cast(target as int)","features")

	val lgbm = new LightGBMClassifier()
	 .setLabelCol("target")
	 .setObjective("binary")
	 .setUseBarrierExecutionMode(true)
	 .setFeaturesCol("features")
	 .setPredictionCol("predictLabel")

	val l: LightGBMClassificationModel =lgbm.fit(df2)
	val predictDF: DataFrame =l.transform(df2.repartition(1))
	predictDF.selectExpr("target","predictLabel").show(false)

My  spark version is 2.4.0, scala version 2.11

**Comment by mbithy** [comment]
not sure if it is related but I have encountered this with a large training set and large boosting and leaves numbers. Lowering the values stopped this error **[LightGBM] [Warning] bad allocation**

**Comment by JWenBin** [comment]
in fact, when my vecCols is less than 5 columns, there is no error occurs. such as:
    val vecCols=Array("country_index","hotid_index","number1","number2")
    val assembler = new VectorAssembler().setInputCols(vecCols).setOutputCol("features")
    ....
only when my vecCols is larger than 5 columns, it seems the more columns my vecCols has, the error occur more frequently

**Comment by JWenBin** [comment]
@mbithy  as you say, with a large training set and more leaves, this error disappeared. But if my training set without "trainDF.limit(N)", there's a new error: BarrierJobUnsupportedRDDChainException. 
Note that "limit(N)" N could be larger than my data or not, it does not matter, but must be set. o(╯□╰)o

**Comment by mbithy** [comment]
check this [Link](https://github.com/Azure/mmlspark/commit/9805996143d4cf174895ff2e08bb61fd2c99c4f1)

**Comment by JWenBin** [comment]
thank you !
